### ISSUE TITLE
Got RuntimeError with circular referenced numbered toctree

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#1536]
I create below as a master_doc::

```
.. toctree::
   :numbered:

   index
```

and I got RuntimeError::

```
# Sphinx version: 1.2.2
# Python version: 2.7.7
# Docutils version: 0.12 release
# Jinja2 version: 2.7.3
# Loaded extensions:
#   sphinx.ext.oldcmarkup from /Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/ext/oldcmarkup.pyc
Traceback (most recent call last):
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/cmdline.py", line 254, in main
    app.build(force_all, filenames)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/application.py", line 212, in build
    self.builder.build_update()
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/builders/__init__.py", line 214, in build_update
    'out of date' &#37; len(to_build))
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/builders/__init__.py", line 244, in build
    for docname in self.env.check_dependents(updated_docnames):
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 490, in check_dependents
    to_rewrite = self.assign_section_numbers()
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1501, in assign_section_numbers
    _walk_toctree(toctreenode, depth)
...
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1490, in _walk_toctree
    self.titles.get(ref))
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1458, in _walk_toc
    _walk_toc(subnode, secnums, depth, titlenode)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1454, in _walk_toc
    _walk_toc(subnode, secnums, depth-1, titlenode)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1478, in _walk_toc
    _walk_toctree(subnode, depth)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1490, in _walk_toctree
    self.titles.get(ref))
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1458, in _walk_toc
    _walk_toc(subnode, secnums, depth, titlenode)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1454, in _walk_toc
    _walk_toc(subnode, secnums, depth-1, titlenode)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1478, in _walk_toc
    _walk_toctree(subnode, depth)
  File "/Users/tkomiya/.buildout/buildout-eggs/Sphinx-1.2.2-py2.7.egg/sphinx/environment.py", line 1483, in _walk_toctree
    for (title, ref) in toctreenode['entries']:
  File "/Users/tkomiya/.buildout/buildout-eggs/docutils-0.12-py2.7.egg/docutils/nodes.py", line 566, in __getitem__
    if isinstance(key, basestring):
RuntimeError: maximum recursion depth exceeded in __instancecheck__
```

If I remove `:numbered:` option from toctree, Sphinx warns my document has a circular toctree reference.

```
/Users/tkomiya/work/tmp/foo/index.rst:: WARNING: circular toctree references detected, ignoring: index <- index
```

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/1536
- Originally reported by: Takeshi Komiya
- Originally created at: 2014-08-18T15:35:05.351

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-08-18 14:27:32+00:00_

Circular toctree is not valid. however it should emits warning instead of raising RuntimeError even if `:numbered:` option is specified as you mentioned.

**Comment by shimizukawa** [comment]
_From Takeshi Komiya on 2014-08-19 02:18:29+00:00_

I send Pull Request #273 . please check it.

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-08-19 21:01:32+00:00_

Merged in tk0miya/sphinx (https://bitbucket.org/birkenfeld/sphinx/pull-request/273)

Fix RuntimeError with numbered circular toctree. Closes #1536

→ <<cset a2a39cefe86665dd3e8c0de18a8982e2ccaba571>>

**Comment by shimizukawa** [comment]
_From Takayuki Shimizukawa on 2014-08-19 21:01:32+00:00_

Merged in tk0miya/sphinx (https://bitbucket.org/birkenfeld/sphinx/pull-request/273)

Fix RuntimeError with numbered circular toctree. Closes #1536

→ <<cset a2a39cefe86665dd3e8c0de18a8982e2ccaba571>>
